File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0, n = prices.size(), i = 1, minLeft = prices[0];
        while(i < n){
            int currProfit = prices[i] - minLeft;
            profit = max(profit, currProfit);
            minLeft = min(minLeft, prices[i]);
            i++;
        }
        return profit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // 7 1 5 3 6 4
        int n = prices.length;
        int maxRightPrice = prices[n-1];
        int maxProfit = 0;
        for(int i = n - 2; i >= 0; i--){
            int currProfit = maxRightPrice - prices[i];
            maxProfit = Math.max(maxProfit, currProfit);
            maxRightPrice = Math.max(maxRightPrice, prices[i]);
        }
        return maxProfit;
    }
}